### PR TITLE
fix(db): Honour an explicitly configured min_conns of 0

### LINKED ISSUE #564: Allow database.minConns: 0 — both the Go config layer and the Helm chart silently coerce it to 2
`database.minConns` cannot be set to `0`. Both the Go config layer and the Helm chart treat an explicit `0` as "unset" and silently substitute `2`, so there is no way to run pgxpool with no warm-connection floor.

This matters for dense multi-tenant deployments, where the per-tenant idle floor is multiplied across every tenant sharing one PostgreSQL server.

## What `minConns: 0` means in pgx/v5

`0` is pgxpool's own default — it is not an unsupported or degenerate value. From `github.com/jackc/pgx/v5@v5.8.0/pgxpool/pool.go`:

```go
defaultMinConns = int32(0)
```

```
// MinConns is the minimum size of the pool. After connection closes, the pool might dip
// below MinConns. A low number of MinConns might mean the pool is empty after
// MaxConnLifetime until the health check has a chance to create new connections.
```

With `MinConns = 0`:

- The pool keeps no warm connections. When traffic stops, connections age out via `MaxConnIdleTime` / `MaxConnLifetime` and the background health check does not recreate them. The pool drains to empty.
- Connections are opened on demand at acquire time.
- `MaxConns` still governs the ceiling, so burst behaviour is unchanged.

The cost is connection setup latency (TCP + TLS + auth + backend fork) on the first query after an idle gap. For workloads that are idle most of the time, that is a good trade.

Harbor's `DefaultMinConns = 2` is therefore a deliberate departure from the library default, and currently a non-overridable one.

## Go config layer

Make "unset" distinguishable from "explicitly 0" — e.g. `*int32` in `models.PostGreSQL`, or a sentinel such as `-1`, or a separate `MinConnsSet bool`. Then:

```go
if cfg.MinConns != nil {
	poolCfg.MinConns = *cfg.MinConns   // honour 0
} else {
	poolCfg.MinConns = DefaultMinConns
}
```

A validation guard rejecting negative values would be worth adding alongside.

## Helm chart

The chart has the same problem independently, so fixing the Go side alone is not enough.

Use an explicit presence check rather than `default`:

```yaml
POSTGRESQL_MIN_CONNS: {{ (hasKey .Values.database "minConns") | ternary .Values.database.minConns 2 | quote }}
```

(or `kindIs "invalid"` on the value, whichever fits the chart's existing conventions).

The `values.schema.json` already permits it — `minConns` is `{"type": "integer", "minimum": 0}` — so the schema expresses an intent the templates do not honour.

## Related: no per-component override

The exporter is the component that least needs a warm pool, but there is no way to give it a different `minConns` from core. `templates/exporter.deployment.yaml` reads the same global `.Values.database.minConns`, and the `exporter.override.database.*` block that existed in the v1 chart is gone.

Even with the two fixes above, tuning the exporter separately would require reinstating a per-component override. Worth considering as part of the same change, though the global fix is the more important half.

## Suggested acceptance criteria

- `POSTGRESQL_MIN_CONNS=0` / `HARBOR_DATABASE_MIN_CONNS=0` results in `pgxpool.Config.MinConns == 0`.
- `database.minConns: 0` in Helm values renders `"0"`, not `"2"`.
- Negative values are rejected with a clear error.

### PR DESCRIPTION
**Pull request by Vad1mo**
## Summary

`database.minConns` could not be set to `0`. The Go config layer treated `0` as "unset" and substituted `DefaultMinConns` (2), so there was no way to run pgxpool without a warm-connection floor. In dense multi-tenant deployments that floor is multiplied across every tenant sharing one PostgreSQL server.

`0` is not a degenerate value — it is pgxpool's own `defaultMinConns`. The pool keeps no warm connections, drains to empty when idle, and opens on demand at acquire time. `MaxConns` still governs the ceiling, so burst behaviour is unchanged. The cost is connection setup latency (TCP + TLS + auth + backend fork) on the first query after an idle gap.

### Why the default was 2

Worth recording, since it looks deliberate and is not. PR #118 (`612a3c5f7`) replaced one metadata line with another in the same hunk:

```go
-{Name: common.PostGreSQLMaxIdleConns, ..., EnvKey: "POSTGRESQL_MAX_IDLE_CONNS", DefaultValue: "2", ...}
+{Name: common.PostGreSQLMinConns,     ..., EnvKey: "POSTGRESQL_MIN_CONNS",     DefaultValue: "2", ...}
```

The `database/sql` knob was a **cap** on idle connections (keep *at most* 2 idle); the pgxpool knob is a **floor** (keep *at least* 2 warm). The number crossed an inverted semantic. The `if x > 0 { } else { default }` guard is the "0 means not set" idiom copied from `MaxOpenConns` directly above it, where 0 genuinely does mean unset for pgxpool.

The default stays 2. It is now overridable.

## Changes

- `models.PostGreSQL.MinConns` becomes `*int32` — nil is "unset", 0 is a configured value.
- `dbpool.applyPoolConfig` honours an explicit 0 and returns an error for a negative value instead of silently falling back. `New` propagates it, so a bad value fails startup rather than quietly changing pool behaviour.
- `ConfigureValue.GetOptionalInt32` returns nil when the key carries no metadata, so a missing key is not read back as a configured 0 (`CfgManager.Get` hands back an empty `ConfigureValue` whose `GetInt` is 0).
- The exporter uses `viper.IsSet`, which distinguishes an unset env var from `HARBOR_DATABASE_MIN_CONNS=0`; `viper.GetInt` collapses both to 0.

## Related Issues

Closes #564

Chart half is on #56 (`fix(chart): Allow database.minConns 0, let exporter.config win over chart env`). Both halves are needed end to end: the chart used to coerce `0` to `2` in its templates independently of this.

## Type of Change

- [x] Bug fix (`fix:`)

## Testing

- [x] Unit tests added/updated

Acceptance criteria from the issue, each asserted directly:

| Criterion | Test |
| --- | --- |
| `MinConns: ptr(0)` → `pgxpool.Config.MinConns == 0` | `TestApplyPoolConfig_ExplicitZeroMinConnsHonored` |
| `MinConns: nil` → 2 | `TestApplyPoolConfig_NilMinConnsUsesDefault` |
| negative rejected with a clear error | `TestApplyPoolConfig_NegativeMinConnsRejected` |
| `POSTGRESQL_MIN_CONNS=0` survives the whole core path | `TestGetDatabaseCfg_MinConns` (config store → `GetDatabaseCfg` → `models.PostGreSQL`) |
| `HARBOR_DATABASE_MIN_CONNS` unset / `""` / `0` / `5` | `TestGetMinConns` |
| an explicit 0 does not read back as unset | `TestConfigureValue_GetOptionalInt32` |

```
go test ./lib/dbpool/... ./lib/config/... ./pkg/config/... ./cmd/exporter/...
ok  github.com/goharbor/harbor/src/lib/dbpool
ok  github.com/goharbor/harbor/src/lib/config/metadata
ok  github.com/goharbor/harbor/src/pkg/config
ok  github.com/goharbor/harbor/src/pkg/config/inmemory
ok  github.com/goharbor/harbor/src/pkg/config/rest
ok  github.com/goharbor/harbor/src/pkg/config/store
ok  github.com/goharbor/harbor/src/cmd/exporter
```

Note for reviewers: five `db`-tagged integration cases set `cfg.MinConns = 0` with a `// don't pre-create` comment, which the old coercion silently turned into 2. They now get what they asked for. `go test -tags db ./lib/dbpool/...` needs a live PostgreSQL and has not been run locally.

## Checklist

- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) format
- [x] Commits are signed off (`git commit -s`)
- [x] No new warnings introduced

### CODE REVIEW

File: src/cmd/exporter/main_test.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			viper.Reset()
+			viper.SetEnvPrefix("harbor")
+			viper.AutomaticEnv()
+			viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+
+			if tt.set {
+				t.Setenv("HARBOR_DATABASE_MIN_CONNS", tt.env)
+			}
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/cmd/exporter/main_test.go:58]: Fixed in 474b630. `t.Setenv` registers restoration of the pre-test value, then the unset case removes the variable — same effect as the save/restore dance, two lines shorter. Also added out-of-range boundary cases while in there.

File: src/lib/config/metadata/value.go [hunk 2]
```diff
@@ -80,6 +80,18 @@ func (c *ConfigureValue) GetInt() int {
...
+// GetOptionalInt32 - return the int32 value of current value, or nil when the
+// value carries no metadata. CfgManager.Get hands back an empty ConfigureValue
+// for an unknown key, which GetInt reports as 0 — indistinguishable from a
+// configured 0. Callers that treat 0 as a meaningful setting use this instead.
+func (c *ConfigureValue) GetOptionalInt32() *int32 {
+	if _, ok := Instance().GetByName(c.Name); !ok {
+		return nil
+	}
+	v := int32(c.GetInt())
+	return &v
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/lib/config/metadata/value.go]: Fixed in 474b630 — and the finding was slightly worse than stated: `parseInt`'s float fallback accepts `"4294967296"` (1<<32), which a bare `int32()` wraps to an **explicit 0** — the opposite of the configured intent, and invisible to the negativity check in `applyPoolConfig`. Both narrowing sites (`GetOptionalInt32`, exporter `getMinConns`) now treat out-of-range as unset with an error/warning log, falling back to the pool default rather than smuggling a wrapped value downstream. Boundary tests use ±1<<32 rather than MaxInt32+1/MinInt32-1: the float32 fallback rounds `-2147483649` to `-2147483648` (exactly representable), so the exact ±1 boundaries don't exercise the range check through the metadata parse path.

File: src/lib/config/metadata/value_test.go [hunk 3]
```diff
@@ -71,6 +72,20 @@ func TestConfigureValue_GetInt(t *testing.T) {
 	assert.Equal(t, createCfgValue("ldap_timeout", "5").GetInt(), 5)
 }
 
+func TestConfigureValue_GetOptionalInt32(t *testing.T) {
+	// A configured 0 must come back as a non-nil 0, not collapse into "unset".
+	zero := createCfgValue("postgresql_min_conns", "0").GetOptionalInt32()
+	require.NotNil(t, zero)
+	assert.Equal(t, int32(0), *zero)
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/lib/config/metadata/value_test.go:84]: Fixed in 474b630 — confirmed real: `TestConfigureValue_GetInt64`/`GetFloat64` swap in `testingMetaDataArray` without restoring, and it lacks `postgresql_min_conns`; the test only passed due to source order. It now pins `Instance().init()` at the start.

File: src/common/models/database.go [hunk 4]
```diff
@@ -37,6 +37,6 @@ type PostGreSQL struct {
 	ConnMaxIdleTime   time.Duration `json:"conn_max_idle_time"`
 	HealthCheckPeriod time.Duration `json:"health_check_period"`
 	ConnectTimeout    time.Duration `json:"connect_timeout"`
-	MinConns          int32         `json:"min_conns"`
+	MinConns          *int32        `json:"min_conns,omitempty"` // nil = unset (dbpool.DefaultMinConns); 0 is a valid setting
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/common/models/database.go:40]: Checked — no wire-format impact. `models.PostGreSQL` is never JSON-serialized: its only consumers are `core/main.go` and `jobservice/job/impl/context.go`, both passing it in-process to `dao.InitDatabase`. The config management API serializes the flat `postgresql_min_conns` metadata item (string-valued store), not this struct; jobservice receives config from core the same way. The `json` tags on this struct are load-bearing only for the harbor.yml→env prepare path in the installer templates, which harbor-next does not use for this model. Intentionally leaving `omitempty` so any incidental marshal reflects "unset" honestly.

File: src/lib/config/metadata/value.go [hunk 5]
```diff
@@ -80,6 +81,26 @@ func (c *ConfigureValue) GetInt() int {
...
+// configured 0. Callers that treat 0 as a meaningful setting use this instead.
+func (c *ConfigureValue) GetOptionalInt32() *int32 {
+	if _, ok := Instance().GetByName(c.Name); !ok {
+		return nil
+	}
+	v := c.GetInt()
+	// parseInt's float fallback admits values beyond int32 (e.g. "4294967296"),
+	// which a bare int32() would wrap — 1<<32 becomes an explicit 0. Treat
+	// out-of-range as unset rather than smuggle a wrapped value downstream.
+	if v < math.MinInt32 || v > math.MaxInt32 {
```
[the remark this answers is not in the document]
Comment (Vad1mo) [reply src/lib/config/metadata/value.go:96]: Declining, deliberately. The premise is off: the exporter does not import `lib/config/metadata` (check `cmd/exporter/main.go` imports), so a helper there would add a new dependency from the exporter binary to the config-metadata package. What the two sites actually share is a two-line range comparison; everything else — log level, message, which key is named — differs on purpose, because each message describes its own parse path (`parseInt` float fallback vs viper `cast.ToInt`). Drift between them is not a risk: each message is about its own site. DRYing this would hide the site-specific comments that explain *why* the guard exists, which is the part worth keeping.

**Comment by Vad1mo** [comment]
#586 does the same thing but a bit more compact. 
duplicate of #586

**Review by bupd — APPROVED** [review]
lgtm
